execution_engine/src/core/engine_state/mod.rs

```diff
@@ -472,6 +472,53 @@ where
                     .ok_or(Error::FailedToGetWithdrawPurses)?
                     .to_owned();
 
+                // Ensure that sufficient balance exists for all unbond purses that are to be
+                // migrated.
+                {
+                    let mut balances = BTreeMap::new();
+                    for purse in withdraw_purses.iter() {
+                        match balances.entry(*purse.bonding_purse()) {
+                            Entry::Vacant(entry) => {
+                                entry.insert(*purse.amount());
+                            }
+                            Entry::Occupied(mut entry) => {
+                                let value = entry.get_mut();
+                                let new_val =
+                                    value.checked_add(*purse.amount()).ok_or_else(|| {
+                                        Error::Mint(
+                                            "overflowed a u512 during unbond migration".into(),
+                                        )
+                                    })?;
+                                *value = new_val;
+                            }
+                        }
+                    }
+
+                    for (unbond_purse_uref, unbond_amount) in balances {
+                        let key = match tracking_copy
+                            .borrow_mut()
+                            .get_purse_balance_key(correlation_id, unbond_purse_uref.into())
+                        {
+                            Ok(key) => key,
+                            Err(_) => return Err(Error::Mint("purse balance not found".into())),
+                        };
+                        let current_balance = tracking_copy
+                            .borrow_mut()
+                            .get_purse_balance(CorrelationId::new(), key)?
+                            .value();
+
+                        if unbond_amount > current_balance {
+                            // If we don't have enough balance to migrate, the only thing we can do
+                            // is to fail the upgrade.
+                            error!("commit_upgrade failed during migration - insufficient funds({current_balance}) in purse({unbond_purse_uref}) to unbond({unbond_amount}) ");
```

Comment:
A nit:
Consider using the tracing values format, like in [here](https://github.com/casper-network/casper-node/blob/dev/execution_engine/src/core/engine_state/mod.rs#L1781).